### ISSUE TITLE
Type narrowing for subclasses is not applied for variable assignments

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#1038]
Bug report from @gvanrossum:

```python
class Animal: ...
class Bear(Animal): ...

def s1() -> Bear:
    a: Animal = Bear()
    # Type of a at this point should be Bear, but it's Animal
   reveal_type(a)

def s2() -> Bear:
    a: Animal
    # Type of a at this point should be Bear, but it's Animal
    a = Bear()
```

As Guido pointed out, Pyright isn't performing subclass-based type narrowing for variable assignments. It does apply subclass-based type narrowing if the original type was a union. It should apply it in any case. This was an oversight, not intended.

Guido is correct when he says "...pyright violates animal rights. :-)"

**Comment by erictraut** [comment]
This bug was introduced several months ago. It's a bad interaction between the type narrowing logic and the "bidirectional type inference" logic that's used for calls to constructors.

I've started to look at improving our regression tests in this area. In doing so, I realized that there are many other cases that we need to consider if we want to standardize narrowing behaviors — cases that involve class hierarchies that involve generics, unions, type arguments that contain Any, unions that contain Any, multiple inheritance, and interesting combinations of the above. I'm going to try to put together a more comprehensive list and a set of principles that we can be applied more generally to cover all of these cases. I won't have much time over the next few days to work on this, but I should get to it later this week or over the weekend.

As a teaser, here are some cases we'll need to consider:

```python
_T1 = TypeVar("_T1")
_T2 = TypeVar("_T2")
_T3 = TypeVar("_T3")

class Animal(Generic[_T1, _T2]):
    ...

class Bear(Animal[_T3, int]):
    ...

class Bat(Animal[int, int]):
    ...

a: Animal[str, int]
a = Bear()
reveal_type(a) # Bear[str]

b: Bear[Any]
b = Bear[int]()
reveal_type(b) # Bear[Any]

c: Animal[Any, Any]
c = Bear[int]()
reveal_type(c) # Bear[Any]

d: Union[Bat, Bear[str]]
d = Bear()
reveal_type(d) # Bear[str]

def f1(e1: Union[Bat, Bear[int]]):
    e: Animal[int, int]
    e = e1
    reveal_type(e) # Union[Bat, Bear[int]]
```

**Comment by gvanrossum** [comment]
Thanks, this will be very helpful!

**Comment by erictraut** [comment]
This is now implemented and will be in the next release of Pyright.

**Comment by erictraut** [comment]
This is now addressed in Pyright 1.1.77, which I just published. It will also be included in the next release of Pylance.
